### PR TITLE
CMake project updates

### PR DESCRIPTION
**Pull request by podsvirov**
A series of improvements:
- Improved  Protobuf module compatibility (disabled by default);
- Hide advanced settings;
- Added build tree configuration;
- Added build of examples.

Dear @xfxyjwf and @yeswalrus please review this changes.

I propose to accept these changes and then continue to improve from @yeswalrus like #1624.
This also replace #1631.

### CODE REVIEW

**Comment by podsvirov** [comment]
Strange ... Travis fell from Python and Ruby on Linux, but it changes only CMake project for Windows.
With appveyor all is good. I think we can ignore Travis and take?
Can anyone clarify the situation?

**Comment by podsvirov** [comment]
Ok. I rebase changes, but appveyor and travis too busy to do a good job for me :-(

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -25,6 +25,13 @@ endif (MSVC)
 option(protobuf_WITH_ZLIB "Build with zlib support" ${protobuf_WITH_ZLIB_DEFAULT})
 set(protobuf_DEBUG_POSTFIX "d"
   CACHE STRING "Default debug postfix")
+mark_as_advanced(protobuf_DEBUG_POSTFIX)
+# User options
+include("${CMAKE_CURRENT_LIST_DIR}/protobuf-options.cmake")
+
+if(protobuf_VERBOSE)
```
Comment (yeswalrus) [inline cmake/CMakeLists.txt]: If you're going to include this type of message, it should go at the top of the file.

File: cmake/CMakeLists.txt [hunk 2]
```diff
@@ -151,3 +158,11 @@ if (protobuf_BUILD_TESTS)
 endif (protobuf_BUILD_TESTS)
 
 include(install.cmake)
+
+if (protobuf_BUILD_EXAMPLES)
```
Comment (yeswalrus) [inline cmake/CMakeLists.txt]: spaces after ifs are non-standard

File: cmake/CMakeLists.txt [hunk 3]
```diff
@@ -151,3 +158,11 @@ if (protobuf_BUILD_TESTS)
...
+  include(examples.cmake)
+endif (protobuf_BUILD_EXAMPLES)
```
Comment (yeswalrus) [inline cmake/CMakeLists.txt]: space after endif is nonstandard, including the if conditional in the endif block should be avoided

File: cmake/install.cmake [hunk 4]
```diff
@@ -6,6 +6,9 @@ foreach(_library
   libprotoc)
   set_property(TARGET ${_library}
     PROPERTY INTERFACE_INCLUDE_DIRECTORIES
+    $<BUILD_INTERFACE:${protobuf_BINARY_DIR}>
```
Comment (yeswalrus) [inline cmake/install.cmake]: Build interface commands should not be included in the install.cmake file

**Comment by yeswalrus** [comment]
@podsvirov https://github.com/google/protobuf/issues/1622. I noticed this too

File: cmake/examples.cmake [hunk 5]
```diff
@@ -0,0 +1,23 @@
+if(protobuf_VERBOSE)
+  message(STATUS "Protocol Buffers Examples Configuring...")
+endif()
+
+# Add examples subproject
+add_subdirectory(../examples examples)
+
+# Add targets for simple invoke example applications
+foreach(TGT add_person list_people)
```
Comment (yeswalrus) [inline cmake/examples.cmake]: These targets are not used anywhere. Remove this foreach.

File: cmake/examples.cmake [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+# Add targets for simple invoke example applications
+foreach(TGT add_person list_people)
+  if(TARGET ${TGT}_cpp)
+    add_custom_target(${TGT}
+      COMMAND ${TGT}_cpp addressbook.pb.bin
+      DEPENDS ${TGT}_cpp
+      WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
+  endif()
+endforeach()
+add_custom_target(clean_people
```
Comment (yeswalrus) [inline cmake/examples.cmake]: again, this target is unused.

File: cmake/install.cmake [hunk 7]
```diff
@@ -87,22 +90,35 @@ if(NOT MSVC)
 else()
   set(CMAKE_INSTALL_CMAKEDIR "cmake" CACHE STRING "${_cmakedir_desc}")
 endif()
+mark_as_advanced(CMAKE_INSTALL_CMAKEDIR)
```
Comment (yeswalrus) [inline cmake/install.cmake]: if debug_postifx & CMAKE_INSTALL_PREFIX aren't advanced variables, this shouldn't be either.

File: cmake/install.cmake [hunk 8]
```diff
@@ -87,22 +90,35 @@ if(NOT MSVC)
...
 
+# Import configuration
 install(EXPORT protobuf-targets
   DESTINATION "${CMAKE_INSTALL_CMAKEDIR}"
   NAMESPACE protobuf::
   COMPONENT protobuf-export)
 
 configure_file(protobuf-config.cmake.in
-  protobuf-config.cmake @ONLY)
+  ${CMAKE_INSTALL_CMAKEDIR}/protobuf-config.cmake @ONLY)
```
Comment (yeswalrus) [inline cmake/install.cmake]: Adding this confuses things. the eventual install directory has nothing to do with a step that happens at configuration time.

File: cmake/install.cmake [hunk 9]
```diff
@@ -87,22 +90,35 @@ if(NOT MSVC)
...
 configure_file(protobuf-config-version.cmake.in
-  protobuf-config-version.cmake @ONLY)
+  ${CMAKE_INSTALL_CMAKEDIR}/protobuf-config-version.cmake @ONLY)
 configure_file(protobuf-module.cmake.in
-  protobuf-module.cmake @ONLY)
+  ${CMAKE_INSTALL_CMAKEDIR}/protobuf-module.cmake @ONLY)
+
+# Build tree import configuration (for import from subprojects)
+if(NOT EXISTS "${protobuf_DIR}" AND NOT EXISTS "${Protobuf_DIR}")
```
Comment (yeswalrus) [inline cmake/install.cmake]: This is not the correct way to support using local builds, and I don't see how it would help when using protobufs as a subproject.

File: cmake/protobuf-options.cmake [hunk 10]
```diff
@@ -0,0 +1,7 @@
+# Verbose output
```
Comment (yeswalrus) [inline cmake/protobuf-options.cmake]: I would suggest removing this file and moving these lines to protobuf-config.in

File: cmake/protobuf-targets.cmake [hunk 11]
```diff
@@ -0,0 +1,17 @@
+# Library aliases
+foreach(_library
```
Comment (yeswalrus) [inline cmake/protobuf-targets.cmake]: Remove this file. protobuf-targets is autogenerated by the install command.

File: cmake/tests.cmake [hunk 12]
```diff
@@ -4,6 +4,7 @@ endif()
 
 option(protobuf_ABSOLUTE_TEST_PLUGIN_PATH
   "Using absolute test_plugin path in tests" ON)
+mark_as_advanced(protobuf_ABSOLUTE_TEST_PLUGIN_PATH)
```
Comment (yeswalrus) [inline cmake/tests.cmake]: this should not be marked as advanced.

File: examples/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,79 @@
+# Minimum CMake required
```
Comment (yeswalrus) [inline examples/CMakeLists.txt]: Adding a CMakeLists.txt file directly in the relevant directory is a good practice, but is counter to the pattern currently used in the rest of the project. @xfxyjwf If you're ok with spreading cmake files out within the rest of the repository, we can do this. Otherwise all this logic should be moved to cmake/examples.cmake

File: examples/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+project(addressbook)
+
+# Find required protobuf package
+find_package(protobuf 3 CONFIG REQUIRED)
+
+if(protobuf_VERBOSE)
+  message(STATUS "Using Protocol Buffers ${Protobuf_VERSION}")
+endif()
+
+file(GLOB SRC_LIST "*.cc")
```
Comment (yeswalrus) [inline examples/CMakeLists.txt]: Never use file(GLOB ...) to gather source lists. It is explicitly discouraged by the documentation for the command.

File: examples/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+file(GLOB PROTO_LIST "*.proto")
```
Comment (yeswalrus) [inline examples/CMakeLists.txt]: as above. bad use of file(GLOB)

File: examples/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+        ${ABS_FIL}
+      DEPENDS ${ABS_FIL}
+      COMMENT "Running C++ protocol buffer compiler on ${FIL}"
+      VERBATIM)
+
+  endforeach()
+
+endif()
+
+include_directories(${CMAKE_CURRENT_BINARY_DIR})
```
Comment (yeswalrus) [inline examples/CMakeLists.txt]: there is a cmake variable you can set to do this automatically.

File: cmake/install.cmake (hunk 8 above)
Comment (podsvirov) [inline cmake/install.cmake]: It is necessary for the subsequent build of the examples as part of the basic process, but examples can also be build separately after the build and installation protobuf (available 2 ways).
Binary folder is similar to the installation folder and has a similar protobuf package configuration files.
Try to enable the option protobuf_BUILD_EXAMPLES.

File: cmake/protobuf-module.cmake.in [hunk 17]
```diff
@@ -1,7 +1,4 @@
-if(PROTOBUF_SRC_ROOT_FOLDER)
-  message(AUTHOR_WARNING "Variable PROTOBUF_SRC_ROOT_FOLDER defined, but not"
-    " used in CONFIG mode")
-endif()
+# Functions
```
Comment (yeswalrus) [inline cmake/protobuf-module.cmake.in]: Instead of essentially copy-pasting FindProtobuf.cmake, couldn't we define the necessary variables and then include it?

File: cmake/protobuf-config.cmake.in [hunk 18]
```diff
@@ -1,19 +1,14 @@
...
 
-# Compute the installation prefix relative to this file.
-get_filename_component(_PROTOBUF_IMPORT_PREFIX
-  "${_PROTOBUF_PACKAGE_PREFIX}" PATH)
-get_filename_component(_PROTOBUF_IMPORT_PREFIX
-  "${_PROTOBUF_IMPORT_PREFIX}" PATH)
-get_filename_component(_PROTOBUF_IMPORT_PREFIX
-  "${_PROTOBUF_IMPORT_PREFIX}" PATH)
+if(protobuf_TARGETS_ALIASES)
+    include("${CMAKE_CURRENT_LIST_DIR}/protobuf-targets-aliases.cmake")
```
Comment (yeswalrus) [inline cmake/protobuf-config.cmake.in]: this file is not included

File: cmake/install.cmake (hunk 4 above)
Comment (podsvirov) [inline cmake/install.cmake]: Why? It is needed for build tree export configuration.

File: cmake/protobuf-config.cmake.in [hunk 19]
```diff
@@ -1,19 +1,14 @@
...
 include("${CMAKE_CURRENT_LIST_DIR}/protobuf-targets.cmake")
 
-# Compute the installation prefix relative to this file.
-get_filename_component(_PROTOBUF_IMPORT_PREFIX
-  "${_PROTOBUF_PACKAGE_PREFIX}" PATH)
-get_filename_component(_PROTOBUF_IMPORT_PREFIX
-  "${_PROTOBUF_IMPORT_PREFIX}" PATH)
-get_filename_component(_PROTOBUF_IMPORT_PREFIX
-  "${_PROTOBUF_IMPORT_PREFIX}" PATH)
+if(protobuf_TARGETS_ALIASES)
```
Comment (yeswalrus) [inline cmake/protobuf-config.cmake.in]: If this is for providing non-namespaced aliases in the config package, I STRONGLY recommend against even providing this as an option.

File: cmake/CMakeLists.txt (hunk 3 above)
Comment (podsvirov) [inline cmake/CMakeLists.txt]: I do not like it, but this style was borrowed from the main list, which was created before me.
I would like to leave it now and work on functionality, and then make corrections of style in all scripts separate PR.

File: cmake/CMakeLists.txt (hunk 2 above)
Comment (podsvirov) [inline cmake/CMakeLists.txt]: I do not like it, but this style was borrowed from the main list, which was created before me.
I would like to leave it now and work on functionality, and then make corrections of style in all scripts separate PR.

File: cmake/protobuf-config.cmake.in (hunk 18 above)
Comment (podsvirov) [inline cmake/protobuf-config.cmake.in]: Now it is removed.

File: cmake/protobuf-config.cmake.in (hunk 19 above)
Comment (podsvirov) [inline cmake/protobuf-config.cmake.in]: Now it is removed.

File: cmake/install.cmake (hunk 7 above)
Comment (podsvirov) [inline cmake/install.cmake]: For example CMAKE_INSTALL_BINDIR and other is advanced variables.
They are used by advanced users and should not interfere with the basic process.
Variable CMAKE_INSTALL_PREFIX know and use all - it does not need to hide in the normal mode.

File: cmake/install.cmake (hunk 9 above)
Comment (podsvirov) [inline cmake/install.cmake]: If variable protobuf_DIR is defined this means that the configuration of this path will be used for the assembly of examples (eg different version protobuf).
Check variable Protobuf_DIR I will remove.

File: cmake/CMakeLists.txt (hunk 2 above)
Comment (yeswalrus) [inline cmake/CMakeLists.txt]: Fair enough, we can leave it for now

File: cmake/protobuf-options.cmake (hunk 10 above)
Comment (podsvirov) [inline cmake/protobuf-options.cmake]: It's a trick :-) I include these file in the main CMakeListst.txt file, and it is used in protobuf-config.cmake.in for all users after installation from package configuration.
This allows these options determine only once.

File: cmake/protobuf-targets.cmake (hunk 11 above)
Comment (podsvirov) [inline cmake/protobuf-targets.cmake]: No for build tree configuration. Needed for build examples in main process (if ptotobuf_BUILD_EXAMPLES enabled).

File: cmake/install.cmake (hunk 4 above)
Comment (yeswalrus) [inline cmake/install.cmake]: the export command is never called though. Adding support for it would be valuable, but should be addressed in a separate PR.

File: cmake/tests.cmake (hunk 12 above)
Comment (podsvirov) [inline cmake/tests.cmake]: Why not? It is only used to build and run tests from the cmake project. This path is not used for the native build process (Linux).
I believe that this is an advanced setting.

File: examples/CMakeLists.txt (hunk 15 above)
Comment (podsvirov) [inline examples/CMakeLists.txt]: What i need use? I need only *.cc files.

File: examples/CMakeLists.txt (hunk 14 above)
Comment (podsvirov) [inline examples/CMakeLists.txt]: What i need use? I need only *.cc files.

File: examples/CMakeLists.txt (hunk 16 above)
Comment (podsvirov) [inline examples/CMakeLists.txt]: You can remind me?

File: cmake/protobuf-options.cmake (hunk 10 above)
Comment (yeswalrus) [inline cmake/protobuf-options.cmake]: gotcha. Makes sense

File: cmake/protobuf-module.cmake.in (hunk 17 above)
Comment (podsvirov) [inline cmake/protobuf-module.cmake.in]: A good idea. In fact, I do. Part of the file after determining the functions of strongly altered to determine the relevant variables, and it is a big part of the file. Simply include the module will not work - you need to check work for all supported versions of CMake. The module had a change over time.

File: cmake/protobuf-targets.cmake (hunk 11 above)
Comment (yeswalrus) [inline cmake/protobuf-targets.cmake]: I believe the recommended way to do this would be with the export() command.

**Comment by podsvirov** [comment]
@yeswalrus I will try to consider your comments and update the code in the near future.

File: cmake/tests.cmake (hunk 12 above)
Comment (yeswalrus) [inline cmake/tests.cmake]: I generally reserve mark_as_advanced for options that are not reccomended, but are user-facing, or for the myriad of <package-name>_* variables that are created by Find modules. I'm fine either way though.

File: examples/CMakeLists.txt (hunk 14 above)
Comment (yeswalrus) [inline examples/CMakeLists.txt]: from https://cmake.org/cmake/help/v3.5/command/file.html :
```
We do not recommend using GLOB to collect a list of source files from your source tree. If no CMakeLists.txt file changes when a source is added or removed then the generated build system cannot know when to ask CMake to regenerate.
```

Comment (yeswalrus) [inline examples/CMakeLists.txt]: it looks like the way protobufs keeps source files in sync is with update_file_lists.sh

File: examples/CMakeLists.txt (hunk 16 above)
Comment (yeswalrus) [inline examples/CMakeLists.txt]: CMAKE_INCLUDE_CURRENT_DIR

File: cmake/install.cmake (hunk 7 above)
Comment (yeswalrus) [inline cmake/install.cmake]: I didn't recall about CMAKE_INSTALL_BINDIR, you're correct. Objection withdrawn.

File: cmake/tests.cmake (hunk 12 above)
Comment (yeswalrus) [inline cmake/tests.cmake]: After thinking some more, I've come around - I think it's probably a good call.

File: cmake/CMakeLists.txt (hunk 1 above)
Comment (podsvirov) [inline cmake/CMakeLists.txt]: Ok. While protobuf_VERBOSE is OFF by default we can move it to top of the file.

**Comment by podsvirov** [comment]
I had the time and I'm working on correcting comments by @yeswalrus .

File: cmake/examples.cmake (hunk 5 above)
Comment (podsvirov) [inline cmake/examples.cmake]: Ok. I remove this.

File: cmake/examples.cmake (hunk 6 above)
Comment (podsvirov) [inline cmake/examples.cmake]: Ok. I remove this.

File: cmake/install.cmake (hunk 4 above)
Comment (podsvirov) [inline cmake/install.cmake]: Ok. I remove this. The remains of experiments with export(EXPORT...

File: cmake/protobuf-targets.cmake (hunk 11 above)
Comment (podsvirov) [inline cmake/protobuf-targets.cmake]: Not because of politics CMP0024. Leave as is - use a ALIASES.

**Comment by podsvirov** [comment]
I finished my work. Changes relocated to the master - waiting for the results from travis and appveyor.
Dear @xfxyjwf , and @yeswalrus , I propose to merge the changes.
The remaining questions regarding the build of examples suggest that to develop in #1643.

**Comment by yeswalrus** [comment]
@xfxyjwf @podsvirov There are still a number of remaining issues here, including the fact that building the examples independently with the protobuf build directory as the protobuf_DIR does not work - #1643 supersedes this and fixes those issues. 

I'm not convinced it's a good idea to support building the examples as part of the master build at all - I agree that there should be a CMakeLists.txt file for them and there should be an option to install the examples directory, but the examples are meant to demonstrate how to use the built and installed protobuf library.

Here's the root of my issue
There are 3 typical usage patterns with CMake:
1. Reference a build and installed package using find_package - This is *the* standard. The other two are not widely supported and create overhead for whoever develops the module
2. Reference a build directory using find_package - This is pretty common too it's what the export command is for.
3. Include the entire project using add_subdirectory - This method has almost nothing in common with the other two. I consider it dangerous, but a few people do it. I consider this to be a bad idea in general. Calling find_package on a project from within itself is extra bad.

What's being done to build the examples in this PR is a mix of 3 and 2, but in such a way that neither quite works independently.

File: cmake/protobuf-targets.cmake (hunk 11 above)
Comment (yeswalrus) [inline cmake/protobuf-targets.cmake]: ALIASES only work when the original targets are also defined. If you do this, when a downstream project's find_package is pointed at the build directory from a project that doesn't include all of protobuf as a sub_directory it will fail terribly. CMP0024 disallows including the result of an export command, but that only happens because we're calling find_package(protobuf...) from *within* protobuf. That function was never meant to do that.

File: cmake/protobuf-module.cmake.in (hunk 17 above)
Comment (yeswalrus) [inline cmake/protobuf-module.cmake.in]: 'Simply include the module will not work'. Not true. I tried this locally and had some promising results.
'The module had a change over time.' This is exactly why I don't like this file.

By including the version of FindProtobuf.cmake that ships with the version of CMake the user has, we guarantee *by definition* that it will be compatible with that version of cmake. If the goal is legacy compatibility you can't get better than that. It also means we'd get any updates for free.

**Comment by yeswalrus** [comment]
https://github.com/google/protobuf/pull/1643 Updated with fixes.
